File: ui/src/main/java/com/xmartlabs/bigbang/ui/common/recyclerview/SingleItemBaseViewHolder.java
package com.xmartlabs.bigbang.ui.common.recyclerview;

import android.support.annotation.CallSuper;
import android.support.annotation.NonNull;
import android.support.annotation.Nullable;
import android.view.View;

import com.annimon.stream.Optional;
import com.xmartlabs.bigbang.core.helper.function.Consumer;

public class SingleItemBaseViewHolder<T> extends BaseViewHolder implements BindingItemViewHolder<T> {
  Optional<Consumer<T>> onClickListener;

  public SingleItemBaseViewHolder(@NonNull View view) {
    this(view, null);
  }

  public SingleItemBaseViewHolder(@NonNull View view, @Nullable Consumer<T> onClickListenerConsumer) {
    super(view);
    this.onClickListener = Optional.ofNullable(onClickListenerConsumer);
  }

  @CallSuper
  @Override
  public void bindItem(@NonNull T item) {
    onClickListener.executeIfPresent(listener ->
        itemView.setOnClickListener(v -> listener.accept(item))
    );
  }
}

File: ui/src/main/java/com/xmartlabs/bigbang/ui/common/recyclerview/SingleItemBaseViewHolder.kt
package com.xmartlabs.bigbang.ui.common.recyclerview

import android.view.View
import androidx.annotation.CallSuper

/**
 * View holder for a single item [T].
 * @param <T> The type of the single item.
 */
open class SingleItemBaseViewHolder<in T> constructor(
    view: View,
    clickListener: ((T) -> Unit)? = null
) : BaseViewHolder(view), BindingItemViewHolder<T> {
  private val onClickListener:  ((T) -> Unit)? = clickListener

  @CallSuper
  override fun bindItem(item: T) {
    onClickListener?.let { listener -> itemView.setOnClickListener { listener(item) } }
  }
}
